plugins/state_monitor.py

```diff
@@ -0,0 +1,194 @@
+        self.low_space = low_space
+        self.sender = None  # Make sure we "initialize" the attribute
+        if method == 'slack':
+            try:
+                self.sender = SlackSender(  # todo how to import?
+                    slack_key_file, slack_channel, logger=self.logger)
+            except FileNotFoundError as e:
+                self.logger.error(
+                    "%e initializing Slack sender: %s Is the Slack key file in the right place?",
+                    type(e).__name__, e)
+                self.logger.info("Error details:", exc_info=True)
+            except Exception as e:  # if anything goes wrong, don't set the sender class
+                self.logger.error("Unexpected %s initializing SlackSender: %s", type(e).__name__, e)
+                self.logger.info("Error details:", exc_info=True)
+
+    def execute(self, input_data=None):
+        """
+        Execute an action upon detection an arbitrary condition in the data.
+
+        Parameters
+        ----------
+        input_data : Mapping[str, DataValue], optional
+            Per iteration input data passed to this function from previous
+            PipelineSteps. Used to extract the data values to report.
+            The default is None.
+
+        Returns
+        -------
+        input_data : same as `input_data`
+            Input data passed through unchanged, for further steps to consume.
+        """
+
+        # Handle first iteration
+        if self._previous_data is None:
+            self._previous_data = input_data
+
+        def now_then(key):
+            # Simple little helper to extract current and previous value for
+            # a single attribute. Variable input_data shadows outer scope
+            return input_data[key].value, self._previous_data[key].value
+
+        try:
+            # Go through several state variables.
+            # If something is hinky, log it and send a message
+
+            # Power drops
+            load_now, load_pre = now_then('adc_vl_f')
+            curr_now, curr_pre = now_then('adc_il_f')
+
+            power_now, power_pre = load_now * curr_now, load_pre * curr_pre
+            if (power_now < self.power_delim) and (power_pre > self.power_delim):
+                msg = f"Power has dropped from {power_pre:.2f} to {power_now:.2f}."
+                self.logger.info(msg)
+                self.send_message(msg)
+
+            # Sensor communication
+            comm_now, comm_pre = now_then('ping')
+            # If the ping !=0, then we can't reach the sensor
+            if comm_now and not comm_pre:
+                msg = "No communication with sensor!"
+                self.logger.info(msg)
+                self.send_message(msg)
+
+            # Remaining triggers
+            space_now, space_pre = now_then('bytes_remaining')
+            if (space_now < self.low_space) and (space_pre > self.low_space):
+                msg = "Remaining GB on drive is %.1f" % space_now
```

Comment:
```suggestion
                msg = f"Remaining GB on drive is {space_now:.1f}"
```

You can use an f-string here, like you do everywhere else